### ISSUE TITLE
compile scylla using C++20 modules

### ISSUE THREAD

**Issue by tchaikov** [scylladb/scylladb#12962]
as a large scale C++ project, it could take long time[0] to compile scylla from scratch. despite that we are using distcc and ccache for speeding up the compilation, the time to build the tree is still sometimes unbearable.

if we split the build process into three phases:

1. prepare the building environment. like installing the dependencies and setting up the environment.
2. create the building system. this is in general where `configure.py` comes into play.
3. generate the source files, and compile the .cc files
4. link the object files into executables

one major contributor of the build time is the 3rd one. where the compiler process large amount the source files. and because current source files are using `#include` preprocess directive to textually include header files recursively, each translation unit could be very large in the sense of its preprocessed source file.

to alleviate the problem, in addition to using distcc and ccache, we could leverage C++20 modules to further shorten the compilation time. it is quite like preprocessed header (PCH) technique found on some compilers, but it is better. as we can define the public interface in a more visible way than using naming convention. and we don't need to split the header from source files for separating the interface and implementation, or for faster compilation.

to build a C++ project using C++20 modules, we need support from following tools

- a capable C++ compiler which does not choke when compiling C++ modules
- a building tool (ninja for instance) which understands the dependency and hence is able to rebuild the outdated source files.
- an optional building system (CMake for instance) which generate the rules consumed by the building tool

---

[0] it takes 8 minutes on a machine with Intel Xeon @ 2.4GHz, 40PU,

**Comment by tchaikov** [comment]
the task can be broken down into following steps:

```[tasklist]
### CMake building system
- [x] add CMake support to scylladb: almost done.
- [ ] verify that CMake and `configure.py` build the same executable by comparing the compiler options and perf-simple-query results.
- [ ] add a switch to `configure.py` to use CMake building system instead its own to generate a top-level `build.ninja`. which can in turn call into the `build.ninja` created by CMake
- [ ] delete old-style building system
```


```[tasklist]
### C++20 module support
- [ ] sort out the dependencies in scylladb to ensure that we don't have cyclic dependencies in different subsystems. we can restructure the building system and code to achieve this goal.
- [ ] ensure our toolchain has decent C++20 module support. see https://github.com/scylladb/seastar/pull/1605
- [ ] reevaluate the support of C++20 module in distcc and ccache
- [ ] reevaluate the support of C++20 module in C++ standard library (this is a nice to have)
- [ ] switch scylladb to C++20 module
```

**Comment by mykaul** [comment]
(nit: you should be able to convert the above comment to a real task list that can be followed)

**Comment by tchaikov** [comment]
it seems that the `[tasklist]` directive only works when in the previewing mode.

**Comment by tchaikov** [comment]
```[tasklist]
- [ ] a
- [ ] b
```

**Comment by regevran** [comment]
If we want to use C++ modules we must make a plan for it.

**Comment by avikivity** [comment]
It's not practical while distcc and ccache don't support it.

It's also not practical while the standard library and the compiler don't match.

**Comment by regevran** [comment]
> It's not practical while distcc and ccache don't support it.

We might be able to get rid of them when we extensively use modules.

> It's also not practical while the standard library and the compiler don't match.

Not sure I understand what match we are missing or hoping to.
The standard library does not support co-routines, compilers do, and we use it.
[Compilers support](https://en.cppreference.com/w/cpp/compiler_support#cpp20).

**Comment by avikivity** [comment]
> > It's not practical while distcc and ccache don't support it.
> 
> We might be able to get rid of them when we extensively use modules.

I don't think so. We'll still have lots of interdependencies. We'll still want switching between branches not to cause a full rebuild.

> 
> > It's also not practical while the standard library and the compiler don't match.
> 
> Not sure I understand what match we are missing or hoping to. The standard library does not support co-routines, compilers do, and we use it. [Compilers support](https://en.cppreference.com/w/cpp/compiler_support#cpp20).

What do coroutines have to do with it?

> 
> As I said, if we want to push in this direction, we need to come up with a plan.

There's no plan we can come up with, the ecosystem is not mature enough.

**Comment by avikivity** [comment]
Absolutely, but it's a mammoth task. I wrote a plan for it somewhere.

**Comment by dani-tweig** [comment]
@regevran is there a plan for how to progress with this one?

**Comment by regevran** [comment]
> @regevran is there a plan for how to progress with this one?

Not really.
We still try to make `configure.py` == `cmake` - once we are OK with this we can continue.

As for C++20 modules (the title of this issue) - there was a prerequisite presented by @avikivity to have `distcc` and `ccache` to have module support. However, building application with modules implies building a [DAG](https://en.wikipedia.org/wiki/Directed_acyclic_graph) staring from the leaves - so no much parallelism is possible, but also nor it is much required.
